src/metabase/models/action.clj

```diff
@@ -352,46 +345,47 @@
 
-(defmethod serdes/extract-query "Action" [_model opts]
-  (eduction (map hydrate-subtype)
-            (t2/reducible-select Action {:where (or (:where opts) true)})))
-
 (defmethod serdes/hash-fields :model/Action [_action]
   [:name (serdes/hydrated-hash :model) :created_at])
 
-(defmethod serdes/extract-one "Action" [_model-name _opts action]
-  (-> (serdes/extract-one-basics "Action" action)
-      (update :creator_id serdes/*export-user*)
-      (update :model_id serdes/*export-fk* 'Card)
-      (update :type name)
-      (cond-> (= (:type action) :query)
-        (update :database_id serdes/*export-fk-keyed* 'Database :name))))
-
-(defmethod serdes/load-xform "Action" [action]
-  (-> action
-      serdes/load-xform-basics
-      (update :creator_id serdes/*import-user*)
-      (update :model_id serdes/*import-fk* 'Card)
-      (update :type keyword)
-      (cond-> (= (:type action) "query")
-        (update :database_id serdes/*import-fk-keyed* 'Database :name))))
-
-(defmethod serdes/ingested-model-columns "Action" [_ingested]
-  (into #{} (conj action-columns :database_id :dataset_query :kind :template :response_handle :error_handle :type)))
-
-(defmethod serdes/load-update! "Action" [_model-name ingested local]
-  (log/tracef "Upserting Action %d: old %s new %s" (:id local) (pr-str local) (pr-str ingested))
-  (update! (assoc ingested :id (:id local)) local)
-  (select-action :id (:id local)))
-
-(defmethod serdes/load-insert! "Action" [_model-name ingested]
-  (log/tracef "Inserting Action: %s" (pr-str ingested))
-  (insert! ingested))
+(defmethod serdes/generate-path "QueryAction" [_ _] nil)
+(defmethod serdes/make-spec "QueryAction" [_model-name _opts]
+  {:copy      []
+   :transform {:action_id     (serdes/parent-ref)
+               :database_id   (serdes/fk :model/Database :name)
+               :dataset_query {:export serdes/export-mbql :import serdes/import-mbql}}})
+
+(defmethod serdes/generate-path "HTTPAction" [_ _] nil)
+(defmethod serdes/make-spec "HTTPAction" [_model-name _opts]
+  {:copy      [:error_handle :response_handle :template]
+   :transform {:action_id (serdes/parent-ref)}})
+
+(defmethod serdes/generate-path "ImplicitAction" [_ _] nil)
+(defmethod serdes/make-spec "ImplicitAction" [_model-name _opts]
+  {:copy      [:kind]
+   :transform {:action_id (serdes/parent-ref)}})
+
+(defmethod serdes/make-spec "Action" [_model-name opts]
+  {:copy      [:archived :created_at :description :entity_id :name :public_uuid :type]
+   :skip      []
+   :transform {:creator_id             (serdes/fk :model/User)
+               :made_public_by_id      (serdes/fk :model/User)
+               :model_id               (serdes/fk :model/Card)
+               :query                  (serdes/nested :model/QueryAction :action_id opts)
+               :http                   (serdes/nested :model/HTTPAction :action_id opts)
+               :implicit               (serdes/nested :model/ImplicitAction :action_id opts)
```

Comment:
Just recording what we discussed for posterity:

1. This is a breaking change to stop merging the "subclass" properties in.
2. Its more verbose and obscures the "meaning" of the relationships a bit.
3. It more transparently tracks the shape of the database (for now).
4. The future of Actions and whether they'll be extended or replaced is unclear.

Reply:
Since this is a breaking change, I'd recommend *not* backporting it.